File: src/main/java/com/wowo/wowo/controllers/UserController.java
package com.wowo.wowo.controllers;

import com.auth0.jwt.interfaces.DecodedJWT;
import com.wowo.wowo.annotations.authorized.IsAdmin;
import com.wowo.wowo.annotations.authorized.IsAuthenticated;
import com.wowo.wowo.data.dto.UserDto;
import com.wowo.wowo.data.mapper.UserMapper;
import com.wowo.wowo.data.mapper.WalletMapper;
import com.wowo.wowo.exceptions.NotFoundException;
import com.wowo.wowo.models.*;
import com.wowo.wowo.repositories.UserRepository;
import com.wowo.wowo.repositories.WalletRepository;
import com.wowo.wowo.services.PartnerService;
import com.wowo.wowo.services.UserService;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.AllArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Optional;

@RestController
@AllArgsConstructor
@RequestMapping(value = "/v1/user", produces = "application/json; charset=UTF-8")
@Tag(name = "User", description = "Người dùng")
@IsAuthenticated
public class UserController {

    private final UserRepository userRepository;
    private final WalletRepository walletRepository;
    private final UserMapper userMapperImpl;
    private final WalletMapper walletMapperImpl;
    private final PartnerService partnerService;
    private final UserService userService;

    @GetMapping()
    @ApiResponse(responseCode = "200", description = "Lấy thông tin thành công",
                 useReturnTypeSchema = true)
    @ApiResponse(responseCode = "404", description = "Không tìm thấy thông tin")
    public Object getUser(Authentication authentication) {
        DecodedJWT decodedJWT = (DecodedJWT) authentication.getDetails();
        String role = decodedJWT.getClaim("role").asString();
        String id = authentication.getPrincipal().toString();
        if (role.contains("partner")) {
            return partnerService.getPartnerById(id).orElseThrow(
                    () -> new NotFoundException("Không tìm thấy thông tin"));
        }
        return userService.getUserById(id);
    }

    @GetMapping("wallet")
    public ResponseEntity<?> getWallet(Authentication authentication) {
        Optional<Wallet> wallet = walletRepository.findByOwnerIdAndOwnerType(
                (String) authentication.getPrincipal(), WalletOwnerType.USER);
        if (wallet.isPresent()) {
            return ResponseEntity.ok(walletMapperImpl.toResponse(wallet.get()));
        }
        return ResponseEntity.notFound().build();
    }

    @GetMapping("/{id}")
    @IsAdmin
    @ApiResponse(responseCode = "200", description = "Lấy thông tin thành công",
                 useReturnTypeSchema = true)
    @ApiResponse(responseCode = "404", description = "Không tìm thấy thông tin")
    @ApiResponse(responseCode = "400", description = "Id không hợp lệ")
    @ApiResponse(responseCode = "403", description = "Không có quyền truy cập")
    public UserDto getUserById(@PathVariable String id) {
        if (id != null) {
            final Optional<User> byId = userRepository.findById(id);
            if (byId.isPresent()) {
                return userMapperImpl.toDto(byId.get());
            }
        }
        return null;
    }


    @GetMapping("/analysis")
    public MonthAnalysis analysis(Authentication authentication) {
        String userId = authentication.getPrincipal().toString();

        return userService.analysis(userId);
    }
}

File: src/main/kotlin/com/wowo/wowo/presentation/rest/UserController.kt
package com.wowo.wowo.presentation.rest

import com.wowo.wowo.contexts.user.application.dto.RegisterUserCommand
import com.wowo.wowo.contexts.user.application.dto.UserDTO
import com.wowo.wowo.contexts.user.application.usecase.GetUserUseCase
import com.wowo.wowo.contexts.user.application.usecase.RegisterUserUseCase
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

/**
 * REST Controller for User operations
 */
@RestController
@RequestMapping("/users")
class UserController(
    private val registerUserUseCase: RegisterUserUseCase,
    private val getUserUseCase: GetUserUseCase
) {

    @PostMapping("/register")
    fun registerUser(@RequestBody request: RegisterUserRequest): ResponseEntity<UserDTO> {
        val command = RegisterUserCommand(
            username = request.username,
            password = request.password,
            email = request.email,
            phoneNumber = request.phoneNumber
        )

        val userDTO = registerUserUseCase.execute(command)
        return ResponseEntity.status(HttpStatus.CREATED).body(userDTO)
    }

    @GetMapping("/{userId}")
    @PreAuthorize("isAuthenticated()")
    fun getUser(@PathVariable userId: String): ResponseEntity<UserDTO> {
        val userDTO = getUserUseCase.execute(userId)
        return ResponseEntity.ok(userDTO)
    }
}

data class RegisterUserRequest(
    val username: String,
    val password: String,
    val email: String,
    val phoneNumber: String?
)

